### PR TITLE
JENKINS-68822 added support for removing all builds with LogRotator

### PR DESCRIPTION
**Pull request by madisparn**
See [JENKINS-68822](https://issues.jenkins.io/browse/JENKINS-68822).

### Testing done
Verified configuration in UI
Checked that old configuration is loaded from disk
Run the unit tests

### Proposed changelog entries

- Allow all builds to be removed by the build discarder.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```
Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/tasks/LogRotator.java
Comment (timja) [inline core/src/main/java/hudson/tasks/LogRotator.java:176]: hmm I think this logic may not be 100% correct.

I have selected "Remove last stable build" - Yes

and then I have a stable build (success) and 2 failures.

My stable build never gets removed because I have set Remove last successful build - no.

The UI seems to be a bit too tied to the implementation. 

At a minimum I think if remove last stable build is selected as yes and the last successful build is a stable build then it should be removed.

----

I wonder if we need both UI options, would it be enough to just add the stable option? 

Or do you have any ideas on how it could be simpler?

Comment (madisparn) [reply core/src/main/java/hudson/tasks/LogRotator.java]: Indeed, having a single UI setting is simpler.

File: core/src/main/resources/hudson/tasks/LogRotator/config.jelly
Comment (timja) [inline core/src/main/resources/hudson/tasks/LogRotator/config.jelly:44]: what about these? The label seems too tied to the developer writing code instead of from a user perspective imo.

```suggestion
    <f:entry title="${%Remove last build}"
```

or 

```suggestion
    <f:entry title="${%Remove last successful build}"
```

successful is still a bit confusing as 'unstable' is successful from my reading of the code.

Comment (madisparn) [reply core/src/main/resources/hudson/tasks/LogRotator/config.jelly]: Picked the shorter option, more clear to me as developer :)

File: core/src/main/java/hudson/tasks/LogRotator.java
Comment (timja) [inline core/src/main/java/hudson/tasks/LogRotator.java:114]:
```suggestion
     * @since TODO
```

**Review by timja — APPROVED** [review]
Thanks!

**Comment by daniel-beck** [comment]
Weak -0.5 as this does not appear to be a common use case and is better off being done in a plugin.

**Review by NotMyFault — APPROVED** [review]

**Comment by NotMyFault** [comment]
/label ready-for-merge

---

This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
Thanks!

**Pull request merged**
